Add blog post: AES67 audio-over-IP on the ESP32-P4 (#746)
espressif/developer-portal: content/blog/2026/06/aes67-audio-over-ip-on-the-esp32-p4/index.md

---
title: "AES67 audio-over-IP on the ESP32-P4"

> date: "2026-06-22"

Reviewer: Please change to `2026-06-24`.

Author: Done — date set to 2026-06-24.

featureAsset: "img/featured/featured-espressif-waves.webp"
authors:
    - sylwester-sosnowski
tags: ["AES67", "RAVENNA", "audio", "PTP", "RTP", "ESP-IDF", "ESP32-P4", "networking"]
---

## What AES67 is, and why this is an odd place to run it

AES67 is an interoperability standard for moving uncompressed audio over a
normal IP network: PCM samples in RTP, sessions described in SDP and
announced over SAP, and — the part that makes it hard — every device locked
to a shared clock by PTP (IEEE-1588) so that streams from different sources
stay sample-aligned. It is what a lot of broadcast trucks, studios, and
live-sound rigs use under the hood, and it interoperates with the RAVENNA
and Dante (AES67 mode) ecosystems.

The hardware that speaks it is usually a dedicated Dante chip, an FPGA, or a
Linux machine with a good NIC. None of those are a microcontroller. The
ESP32-P4 is interesting here because it has two things that make AES67

> plausible on a part this size: a RISC-V core fast enough to convert and move

Reviewer 2: ```suggestion
plausible on a device of this size: a RISC-V core fast enough to convert and move
```

Author: Applied, thanks.

audio samples in software, and an Ethernet MAC with **IEEE-1588 hardware
timestamping** — which is the one feature you cannot fake if you want real
PTP sync.

> So I built an AES67/RAVENNA endpoint as an ESP-IDF component for the P4. It

Reviewer: P4 -> ESP32-P4

Author: Done.

synchronizes to an external PTP grandmaster (or becomes one), sends and
receives multichannel RTP audio, discovers and is discovered over SAP/SDP,
and plays out through I2S to a DAC. It interoperates on real hardware with
Merging Technologies SIENNA, with the
[aes67-linux-daemon](https://github.com/bondagit/aes67-linux-daemon) running
on a Raspberry Pi, and with a standalone AES67 hardware speaker/amplifier.
End-to-end latency, best case, is about **0.7 ms**.

This article walks through the three parts that actually matter — the clock,
getting packets off the wire fast, and playing them out without jitter — and
is honest about what's solid and what isn't.

## The clock is the whole problem

> In most network audio the headline number people quote is latency, but the

Reviewer 2: ```suggestion
When discussing network audio, latency is the key metric people quote. But the
```

Author: Applied, thanks.

A couple of details worth pulling out:

- **The PTP identity comes from the MAC address.** A PTP clock needs a
  unique 64-bit identity; AES67 devices derive it from the 48-bit MAC by the
  standard EUI-64 expansion (insert `FF:FE` in the middle). The node also
  uses that to detect when *it* has been elected grandmaster — it compares
  the announced grandmaster ID against its own MAC-derived identity.
- **It can be master or slave.** If there's a better clock on the network,
  the node locks to it. If there isn't, best-master-clock selection promotes
  this node to grandmaster and it serves time to everyone else. That's not a
  mode switch you configure; it falls out of the protocol.

Once the clock is disciplined, RTP timestamps are just a projection of PTP
time onto the media rate: `rtp_ts = ptp_time_ns * sample_rate / 1e9`. Get
the clock right and the rest of the timing is arithmetic.

## Getting RTP off the wire without paying for the network stack

The receive path is where the latency budget is won or lost. A normal
sockets path — EMAC interrupt, into lwIP, IP/UDP demux, copy into a socket
buffer, wake the reader task — adds buffering and scheduling delay at every

> hop, and on a part this size that's a meaningful fraction of a millisecond

Reviewer 2: ```suggestion
hop, and on a device of this size that's a meaningful fraction of a millisecond
```

Author: Applied, thanks.

    if ((buf[14 + 16] & 0xF0) != 0xE0) goto forward;  /* not multicast       */
    /* ... port 5004? RTP v2? then parse the header and convert the payload  */
    /* straight into the audio stream buffer, and return — lwIP never sees   */
    /* this frame. */
forward:
    return s_original_input(eth, buf, len, priv, info);  /* back to lwIP     */
}
```

The checks are ordered cheapest-first and the whole thing lives in IRAM so
it isn't waiting on flash. A non-audio frame pays only a handful of byte
comparisons before being handed back to lwIP, so the rest of the network
stack — DHCP, the web UI, mDNS — keeps working normally. An audio frame
never enters lwIP at all: its payload is converted (L16/L24/L32 to the
internal int32 format) and written to the playout buffer right there in the
driver callback.

To be precise about what's where: the L2-tap receive path is in the
component and portable; the in-driver hook above is example code in the
repository's `main/`, because it's tied to how you install your Ethernet
driver. It's the path that produces the lowest numbers, but it's wiring you
copy and adapt, not a component API.

## Playing it out without jitter

> On the playback side the enemy is the same one in a different costume:

Reviewer 2: ```suggestion
On the playback side the enemy is the same one but wearing different clothes:
```

Author: Changes applied

Author: Applied, thanks.

The other thing that didn't work until it did was multicast under load. The
ESP32-P4 EMAC would drop multicast frames once the packet rate got high
enough — about 8% loss — which on an audio stream is constant clicking.
Enabling **IEEE 802.3x flow control** so the MAC can pause the sender when
its buffers fill fixed it: zero loss across a 645,000-packet test, no
sequence gaps. Flow control is one Kconfig option and a line in the driver
config, and it's the difference between "demo" and "leave it running."

## Does it actually interoperate

This is the part that separates an AES67 implementation from a thing that
sends RTP packets. The standard exists so that gear from different vendors
locks together, so the only test that counts is against other vendors' gear.

I ran it against three reference points. The first is the
[aes67-linux-daemon](https://github.com/bondagit/aes67-linux-daemon) on a
Raspberry Pi — an open-source AES67 implementation — as both a source and a
sink, with the P4 discovering it over SAP and locking to its PTP. The second
is **Merging Technologies SIENNA**, a commercial RAVENNA implementation; the

> P4 discovered its announced streams, parsed its SDP, synchronized to the

Reviewer: P4 -> ESP32-P4
Check for any other missing ESP32-.

Author: Done — swept the whole article and replaced all 6 standalone 'P4' with 'ESP32-P4'.
